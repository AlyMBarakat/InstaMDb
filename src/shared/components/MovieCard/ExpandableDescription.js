import React, { useState, useCallback } from 'react';
import {
    View,
    TouchableOpacity,
    LayoutAnimation,
} from 'react-native';
import Text from '../Text';

const ExpandableDescription = ({ text }) => {
    const [seeMoreButton, setSeeMoreButton] = useState(false); // see more button visibility
    const [textEnlarged, setTextEnlarged] = useState(false);
    const MAX_NUM_OF_LINES = 4; // maximum lines to be displayed while collapsed

    // toggled to alternate between see more and see less
    const toggleDescriptionLength = () => {
        // animate components transitions
        LayoutAnimation.configureNext(
            LayoutAnimation.create(
                150,
                LayoutAnimation.Types.linear,
                LayoutAnimation.Properties.opacity,
            )
        );
        setTextEnlarged(!textEnlarged);
    }
    // Determine if see more button should be displayed or not
    // based on number of lines returned by text layout change event
    const onTextLayout = useCallback(
        textLayoutEvent => {
            setSeeMoreButton(textLayoutEvent.nativeEvent.lines.length >= MAX_NUM_OF_LINES);
        }, []);

    return (
        <View>
            {/* Main description text */}
            <Text
                light
                size={14}
                color='#988989'
                // zero is the default value for numberOfLines ~ show all text
                numberOfLines={textEnlarged ? 0 : MAX_NUM_OF_LINES}
                selectable={true}
                onTextLayout={onTextLayout}
                style={{ lineHeight: 18 }}
            >
                {text}
            </Text >
            {/* See more / See less button */}
            {seeMoreButton &&
                <TouchableOpacity
                    style={{ alignSelf: 'flex-end' }}
                    onPress={toggleDescriptionLength}
                >
                    <Text light
                        size={14}
                        color='#fff'
                    >{textEnlarged ? "See less" : "See more"}</Text>
                </TouchableOpacity >
            }
        </View>
    );
}
export default ExpandableDescription;