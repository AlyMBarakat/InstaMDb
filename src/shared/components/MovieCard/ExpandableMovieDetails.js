import React, { useState, useCallback } from 'react';
import {
    View,
    TouchableOpacity,
    LayoutAnimation,
} from 'react-native';
import Text from '../Text';

const ExpandableDescription = ({ title, releaseDate, overview }) => {
    const [seeMoreButton, setSeeMoreButton] = useState(false); // see more button visibility
    const [textEnlarged, setTextEnlarged] = useState(false);
    const MAX_NUM_OF_LINES = 4; // maximum lines to be displayed while collapsed

    // toggled to alternate between see more and see less
    const toggleDescriptionLength = () => {
        // animate components transitions
        LayoutAnimation.configureNext(
            LayoutAnimation.create(
                300,
                LayoutAnimation.Types.linear,
                LayoutAnimation.Properties.opacity,
            )
        );
        setTextEnlarged(!textEnlarged);
    }
    // Determine if see more button should be displayed or not
    // based on number of lines returned by text layout change event
    const onTextLayout = useCallback(
        textLayoutEvent => {
            setSeeMoreButton(textLayoutEvent.nativeEvent.lines.length >= MAX_NUM_OF_LINES);
        }, []);

    return (
        <View>
            {/* Title */}
            <Text bold size={18} color='#D73B3B'>{title}</Text>
            {/* Release Date */}
            <Text size={13} color='#D88282' style={{ marginVertical: 3 }}>{releaseDate}</Text>
            {/* Main overview text */}
            <Text
                light
                size={14}
                color='#988989'
                // zero is the default value for numberOfLines ~ show all text
                numberOfLines={textEnlarged ? 0 : MAX_NUM_OF_LINES}
                selectable={true}
                onTextLayout={onTextLayout}
                style={{ lineHeight: 18 }}
            >
                {overview}
            </Text >
            {/* See more / See less button */}
            {seeMoreButton &&
                <TouchableOpacity
                    style={{ alignSelf: 'flex-end', padding: 5 }}
                    onPress={toggleDescriptionLength}
                >
                    <Text medium
                        size={14}
                        color='#fff'
                    >{textEnlarged ? "See less" : "See more"}</Text>
                </TouchableOpacity >
            }
        </View>
    );
}
export default ExpandableDescription;