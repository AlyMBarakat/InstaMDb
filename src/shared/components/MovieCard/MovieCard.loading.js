import React from "react";
import { View } from "react-native";
import SkeletonPlaceholder from "react-native-skeleton-placeholder";
import styles from './MovieCard.loading.styles';

const MovieCardLoading = () => {
    return (
        <SkeletonPlaceholder backgroundColor="#351818" highlightColor="#A71E1E" >
            <View style={styles.container}>
                <View style={styles.poster}>
                    <View style={{ flex: 1 }} />
                </View>
                <View style={styles.textContainer}>
                    <View style={[{ width: "100%", height: 20 }, styles.text]} />
                    <View
                        style={[{ width: "40%", height: 20 }, styles.text]}
                    />
                    <View
                        style={[{ width: "100%", height: 50 }, styles.text]}
                    />
                </View>
            </View>
        </SkeletonPlaceholder>
    );
};

export default MovieCardLoading;
