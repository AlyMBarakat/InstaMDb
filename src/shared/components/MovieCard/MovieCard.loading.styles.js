import { StyleSheet } from 'react-native';

const styles = new StyleSheet.create({
    container: {
        borderRadius: 5,
        flexDirection: 'row',
        justifyContent: 'flex-start',
        marginBottom: 10,
    },
    textContainer: {
        flex: 1,
        justifyContent: 'center',
        marginHorizontal: 10,
        marginVertical: 10,
    },
    poster: {
        width: "40%",
        aspectRatio: 2 / 3,
        borderTopLeftRadius: 5,
        borderBottomLeftRadius: 5,
        backgroundColor: '#351818',
    },
    text: {
        marginTop: 6,
        borderRadius: 4,
        backgroundColor: 'red',
    },

});

export default styles;