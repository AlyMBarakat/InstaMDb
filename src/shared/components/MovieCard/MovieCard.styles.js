import { StyleSheet } from 'react-native';

const styles = new StyleSheet.create({
    container: {
        backgroundColor: '#351818',
        borderRadius: 5,
        flexDirection: 'row',
        justifyContent: 'flex-start',
    },
    poster: {
        backgroundColor: 'white',
        width: "40%",
        aspectRatio: 2 / 3,
        borderTopLeftRadius: 5,
        borderBottomLeftRadius: 5,
    },
    textContainer: {
        flex: 1,
        justifyContent: 'center',
        marginHorizontal: 10,
        marginVertical: 10,
    },
});

export default styles;