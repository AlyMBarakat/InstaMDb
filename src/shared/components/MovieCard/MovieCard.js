import React from 'react';
import { View, Image } from 'react-native';
import Text from '../Text';
import styles from './MovieCard.styles';
import ExpandableDescription from './ExpandableDescription';

const MovieCard = React.memo(({ movieDetails, style }) => {
    return (
        <View style={[style, styles.container]}>
            <Image
                source={{ uri: `https://image.tmdb.org/t/p/w342/${movieDetails.posterPath}` }}
                style={styles.poster}
            />
            {/* Movie's data */}
            <View style={styles.textContainer}>
                {/* Title */}
                <Text bold size={18} color='#D73B3B'>{movieDetails.title}</Text>
                {/* Date */}
                <Text size={13} color='#D88282' style={{ marginVertical: 3 }}>{movieDetails.releaseDate}</Text>
                {/* ExpandableDescription */}
                <ExpandableDescription text={movieDetails.overview} />
            </View >
        </View >
    );
});

export default MovieCard;