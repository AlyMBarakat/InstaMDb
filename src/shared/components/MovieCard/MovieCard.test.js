import React from 'react';
import renderer from 'react-test-renderer';
import MovieCard from '../MovieCard';

const movieDetailsDummy = { "id": 726684, "overview": "To join Adrien in Shanghai, Marinette is going to visit her uncle Wang who is celebrating his anniversary. But, as soon as she arrives in China, her purse gets stolen with Tikki inside, whom she needs to secretly transform into Ladybug! Without money and alone in the immense city, Marinette accepts the help of a young and resourceful girl, Fei. The two girls will ally and discover the existence of a new magical jewel, the Prodigious. Hawk Moth, also present in Shanghai, seeks to finding it since a long time...", "posterPath": "/msI5a9TPnepx47JUb2vl88hb80R.jpg", "releaseDate": "2021-04-04", "title": "Miraculous World: Shanghai â€“ The Legend of Ladydragon" }


test('renders movie card loading skeleton correctly', () => {
    const tree = renderer.create(<MovieCard loading />).toJSON();
    expect(tree).toMatchSnapshot();
});

test('renders movie card correctly', () => {
    const tree = renderer.create(<MovieCard movieDetails={movieDetailsDummy} />).toJSON();
    expect(tree).toMatchSnapshot();
});
