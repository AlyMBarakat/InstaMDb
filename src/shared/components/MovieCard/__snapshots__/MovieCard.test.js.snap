// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  onLayout={[Function]}
>
  <View
    style={
      Object {
        "borderRadius": 5,
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "marginBottom": 10,
      }
    }
  >
    <View
      style={
        Object {
          "aspectRatio": 0.6666666666666666,
          "backgroundColor": "#351818",
          "borderBottomLeftRadius": 5,
          "borderTopLeftRadius": 5,
          "width": "40%",
        }
      }
    >
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
              },
              Object {
                "backgroundColor": "#351818",
                "overflow": "hidden",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
          "marginHorizontal": 10,
          "marginVertical": 10,
        }
      }
    >
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Array [
                Object {
                  "height": 20,
                  "width": "100%",
                },
                Object {
                  "backgroundColor": "red",
                  "borderRadius": 4,
                  "marginTop": 6,
                },
              ],
              Object {
                "backgroundColor": "#351818",
                "overflow": "hidden",
              },
            ]
          }
        />
      </View>
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Array [
                Object {
                  "height": 20,
                  "width": "40%",
                },
                Object {
                  "backgroundColor": "red",
                  "borderRadius": 4,
                  "marginTop": 6,
                },
              ],
              Object {
                "backgroundColor": "#351818",
                "overflow": "hidden",
              },
            ]
          }
        />
      </View>
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Array [
                Object {
                  "height": 50,
                  "width": "100%",
                },
                Object {
                  "backgroundColor": "red",
                  "borderRadius": 4,
                  "marginTop": 6,
                },
              ],
              Object {
                "backgroundColor": "#351818",
                "overflow": "hidden",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly 2`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "backgroundColor": "#351818",
        "borderRadius": 5,
        "flexDirection": "row",
        "justifyContent": "flex-start",
      },
    ]
  }
>
  <Image
    source={
      Object {
        "uri": "https://image.tmdb.org/t/p/w342//msI5a9TPnepx47JUb2vl88hb80R.jpg",
      }
    }
    style={
      Object {
        "aspectRatio": 0.6666666666666666,
        "borderBottomLeftRadius": 5,
        "borderTopLeftRadius": 5,
        "width": "40%",
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 10,
        "marginVertical": 10,
      }
    }
  >
    <View>
      <Text
        style={
          Array [
            Object {
              "color": "#D73B3B",
              "fontFamily": "Rubik-Bold",
              "fontSize": 18,
            },
            undefined,
          ]
        }
      >
        Miraculous World: Shanghai – The Legend of Ladydragon
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#D88282",
              "fontFamily": "Rubik-Medium",
              "fontSize": 13,
            },
            Object {
              "marginVertical": 3,
            },
          ]
        }
      >
        2021-04-04
      </Text>
      <Text
        numberOfLines={4}
        onTextLayout={[Function]}
        selectable={true}
        style={
          Array [
            Object {
              "color": "#988989",
              "fontFamily": "Rubik-Light",
              "fontSize": 14,
            },
            Object {
              "lineHeight": 18,
            },
          ]
        }
      >
        To join Adrien in Shanghai, Marinette is going to visit her uncle Wang who is celebrating his anniversary. But, as soon as she arrives in China, her purse gets stolen with Tikki inside, whom she needs to secretly transform into Ladybug! Without money and alone in the immense city, Marinette accepts the help of a young and resourceful girl, Fei. The two girls will ally and discover the existence of a new magical jewel, the Prodigious. Hawk Moth, also present in Shanghai, seeks to finding it since a long time...
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders movie card correctly 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "backgroundColor": "#351818",
        "borderRadius": 5,
        "flexDirection": "row",
        "justifyContent": "flex-start",
      },
    ]
  }
>
  <Image
    source={
      Object {
        "uri": "https://image.tmdb.org/t/p/w342//msI5a9TPnepx47JUb2vl88hb80R.jpg",
      }
    }
    style={
      Object {
        "aspectRatio": 0.6666666666666666,
        "borderBottomLeftRadius": 5,
        "borderTopLeftRadius": 5,
        "width": "40%",
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 10,
        "marginVertical": 10,
      }
    }
  >
    <View>
      <Text
        style={
          Array [
            Object {
              "color": "#D73B3B",
              "fontFamily": "Rubik-Bold",
              "fontSize": 18,
            },
            undefined,
          ]
        }
      >
        Miraculous World: Shanghai – The Legend of Ladydragon
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#D88282",
              "fontFamily": "Rubik-Medium",
              "fontSize": 13,
            },
            Object {
              "marginVertical": 3,
            },
          ]
        }
      >
        2021-04-04
      </Text>
      <Text
        numberOfLines={4}
        onTextLayout={[Function]}
        selectable={true}
        style={
          Array [
            Object {
              "color": "#988989",
              "fontFamily": "Rubik-Light",
              "fontSize": 14,
            },
            Object {
              "lineHeight": 18,
            },
          ]
        }
      >
        To join Adrien in Shanghai, Marinette is going to visit her uncle Wang who is celebrating his anniversary. But, as soon as she arrives in China, her purse gets stolen with Tikki inside, whom she needs to secretly transform into Ladybug! Without money and alone in the immense city, Marinette accepts the help of a young and resourceful girl, Fei. The two girls will ally and discover the existence of a new magical jewel, the Prodigious. Hawk Moth, also present in Shanghai, seeks to finding it since a long time...
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders movie card loading skeleton correctly 1`] = `
<View
  onLayout={[Function]}
>
  <View
    style={
      Object {
        "borderRadius": 5,
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "marginBottom": 10,
      }
    }
  >
    <View
      style={
        Object {
          "aspectRatio": 0.6666666666666666,
          "backgroundColor": "#351818",
          "borderBottomLeftRadius": 5,
          "borderTopLeftRadius": 5,
          "width": "40%",
        }
      }
    >
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
              },
              Object {
                "backgroundColor": "#351818",
                "overflow": "hidden",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
          "marginHorizontal": 10,
          "marginVertical": 10,
        }
      }
    >
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Array [
                Object {
                  "height": 20,
                  "width": "100%",
                },
                Object {
                  "backgroundColor": "red",
                  "borderRadius": 4,
                  "marginTop": 6,
                },
              ],
              Object {
                "backgroundColor": "#351818",
                "overflow": "hidden",
              },
            ]
          }
        />
      </View>
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Array [
                Object {
                  "height": 20,
                  "width": "40%",
                },
                Object {
                  "backgroundColor": "red",
                  "borderRadius": 4,
                  "marginTop": 6,
                },
              ],
              Object {
                "backgroundColor": "#351818",
                "overflow": "hidden",
              },
            ]
          }
        />
      </View>
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Array [
                Object {
                  "height": 50,
                  "width": "100%",
                },
                Object {
                  "backgroundColor": "red",
                  "borderRadius": 4,
                  "marginTop": 6,
                },
              ],
              Object {
                "backgroundColor": "#351818",
                "overflow": "hidden",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
